src/main/java/io/aiven/kafka/auth/AivenAclAuthorizer.java

```diff
@@ -80,41 +75,58 @@ public void configure(final java.util.Map<String, ?> configs) {
 
         configFile = config.getConfigFile();
         jsonReader = new AclJsonReader(configFile.toPath());
-        checkAndUpdateConfig();
+        reloadAcls();
+        final WatchKey watchKey = subscribeToAclChanges(configFile);
+        scheduledFuture = scheduledExecutorService.scheduleWithFixedDelay(() -> {
+            final List<WatchEvent<?>> watchEvents = watchKey.pollEvents();
+            watchEvents.stream().filter(watchEvent -> {
+                @SuppressWarnings("unchecked")
+                final Path path = ((WatchEvent<Path>) watchEvent).context();
+                return path.endsWith(configFile.getName());
+            }).findFirst().ifPresent(watchEvent -> {
+                LOGGER.info(
+                        watchEvent.kind() + ": " + watchEvent.context() + ", Modified: " + configFile.lastModified());
+                reloadAcls();
+            });
+        }, 0, config.configRefreshTimeout(), TimeUnit.MILLISECONDS);
+    }
+
+    private WatchKey subscribeToAclChanges(final File configFile) {
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            return configFile.toPath().getParent().register(watchService, StandardWatchEventKinds.ENTRY_MODIFY);
+        } catch (final IOException e) {
+            LOGGER.error("Failed to subscribe to ACL configuration changes");
+            throw new RuntimeException("Failed to subscribe to ACL configuration changes", e);
+        }
     }
 
     /**
      * Read config file and populate ACL entries, if the config file has changed.
      * This function assumes appropriate synchronization by caller.
      */
-    private void checkAndUpdateConfig() {
-        final long configFileLastModified = configFile.lastModified();
-
-        if (configFileLastModified != lastModifiedTimestamp) {
-            LOGGER.info("Reloading ACL configuration {}", configFile);
-            try {
-                final List<AivenAcl> newAclEntries = jsonReader.read();
-                // initialize cache for non-trivial ACLs
-                if (newAclEntries.size() > 10) {
-                    if (verdictCache != null) {
-                        verdictCache.clear();
-                    } else {
-                        verdictCache = new ConcurrentHashMap<>();
-                    }
-                } else {
-                    verdictCache = null;
-                }
-                aclEntries = newAclEntries;
-            } catch (final JsonReaderException ex) {
-                LOGGER.error("Failed to load config file", ex);
-            }
+    public boolean reloadAcls() {
+        LOGGER.info("Reloading ACL configuration {}", configFile);
+        try {
+            final List<AivenAcl> newAclEntries = jsonReader.read();
+            verdictCache.clear();
+            aclEntries.clear();
+            return aclEntries.addAll(newAclEntries);
```

Comment:
The calls to `aclEntries.clear();` and `aclEntries.addAll(newAclEntries);` are not atomic. I think it is possible to have the operations interleaved like this:


| watcher | process thread
| --- | ---
| clear() |
| | aclEntries.stream()
| addAll() |

I'm not sure about the implementation details here, but that could result in a valid request being reject until the configuration file is modified.